### PR TITLE
Allow destination shape to be specified in reproject()

### PR DESCRIPTION
**Pull request by davidbrochart**
- [x] Tests added
 - [x] Fully documented, including `docs/history.rst` for all changes and `docs/rioxarray.rst` for new API

### CODE REVIEW

File: rioxarray/rioxarray.py [hunk 1]
```diff
@@ -871,6 +885,8 @@ def reproject(
         resolution: float or tuple(float, float), optional
             Size of a destination pixel in destination projection units
             (e.g. degrees or metres).
+        shape: tuple(int, int), optional
+            Shape of the destination in pixels. Cannot be used together with resolution.
```
Comment (snowman2) [inline rioxarray/rioxarray.py:889]: Mind adding a note that the order in the tuple is (dst_height, dst_width)?

**Comment by snowman2** [comment]
@davidbrochart, this looks great, thanks :+1:. Just a minor comment above and I think that it would be good to add another test that ensures the output shape is expected when using the `shape` argument. Would you mind adding that in?

**Comment by snowman2** [comment]
Also, just thought of this - mind adding `.. versionadded:: 0.0.27 shape` to above the parameters section in the docstrings of both `reproject` methods?

File: rioxarray/rioxarray.py (hunk 1 above)
Comment (davidbrochart) [reply rioxarray/rioxarray.py]: :+1:

File: rioxarray/rioxarray.py [hunk 2]
```diff
@@ -203,12 +203,29 @@ def _make_coords(src_data_array, dst_affine, dst_width, dst_height, dst_crs):
...
 
-def _make_dst_affine(src_data_array, src_crs, dst_crs, dst_resolution=None):
+def _make_dst_affine(
+    src_data_array, src_crs, dst_crs, dst_resolution=None, dst_shape=None
+):
     """Determine the affine of the new projected `xarray.DataArray`"""
     src_bounds = src_data_array.rio.bounds()
     src_width, src_height = src_data_array.rio.shape
+    dst_height, dst_width = dst_shape if dst_shape is not None else (None, None)
+    resolution_or_width_height = {
```
Comment (davidbrochart) [inline rioxarray/rioxarray.py:213]: Should `shape` be `(dst_width, dst_height)`, to conform to the source shape?

Comment (snowman2) [reply rioxarray/rioxarray.py:213]: If you explicitly pass in `dst_width` `dst_height`, rasterio will do the validation for you:
https://github.com/mapbox/rasterio/blob/ce00d482faa24068f41164b79901eeec31777263/rasterio/warp.py#L457-L467

Comment (snowman2) [reply rioxarray/rioxarray.py:213]: > Should shape be (dst_width, dst_height), to conform to the source shape?

That is actually a bug I think - those should be the other way around.

Comment (snowman2) [reply rioxarray/rioxarray.py:213]: > should we explicitly pass dst_width and dst_height?

I am fine either way. I like `shape` but `dst_width` and `dst_height` are used by `rasterio` and make the code simpler as you just pass in the values.

Comment (davidbrochart) [reply rioxarray/rioxarray.py:213]: I was also thinking that rioxarray offers a somewhat higher-level API than rasterio, and more similar to NumPy. That's why I thought that `(dst_height, dst_width)` is really the shape of the array. So I'm fine with keeping things like that, if you tell me that the source shape should also be `(src_height, src_width)`.

Comment (snowman2) [reply rioxarray/rioxarray.py:213]: Sounds good, then I think it should be good to go as is.

**Comment by snowman2** [comment]
@davidbrochart did you want to squash your commits with me doing a merge commit or would you rather I just do a squash-and-merge?

**Comment by davidbrochart** [comment]
Yes, please go ahead and do a squash-and-merge.

**Comment by snowman2** [comment]
Thanks @davidbrochart for this addition :+1:

**Comment by davidbrochart** [comment]
Thanks a lot for this awesome library!

**Pull request merged**